AutoDarkModeSvc/Modules/ProcessBlockList.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System.Diagnostics;
+using System.Linq;
+using AutoDarkModeLib;
+using AutoDarkModeSvc.Core;
+using AutoDarkModeSvc.Timers;
+
+namespace AutoDarkModeSvc.Modules;
+
+public class ProcessBlockList : AutoDarkModeModule
+{
+    private static readonly NLog.Logger Logger = NLog.LogManager.GetCurrentClassLogger();
+    private GlobalState State { get; }
+    private AdmConfigBuilder ConfigBuilder { get; }
+
+    public ProcessBlockList(string name, bool fireOnRegistration) : base(name, fireOnRegistration)
+    {
+        State = GlobalState.Instance();
+        ConfigBuilder = AdmConfigBuilder.Instance();
+    }
+
+    private const string ExcludedProcessIsRunningReason = "Excluded Process is running";
+
+    public override string TimerAffinity { get; } = TimerName.Main;
+
+    public override void Fire()
+    {
+        if (ConfigBuilder.Config.ProcessBlockList.ProcessNames.Count == 0)
+        {
+            Logger.Info("No processes are excluded, skipping checks");
+            return;
+        }
+
+        var activeProcesses = Process
+            .GetProcesses()
+            .Where(p => p.MainWindowHandle != 0)
+            .Select(p => p.ProcessName)
+            .ToHashSet();
+
+        if (ConfigBuilder.Config.ProcessBlockList.ProcessNames.Any(p => activeProcesses.Contains(p)))
```

Comment:
Detection logic shouldn't run constantly.

Similar to the GPUMonitorModule, it should start monitoring a minute or two (for the GPU monitor module that value is customizable) before a theme switch would occur, then hold and monitor for as long as necessary. After the program that causes the delay is closed, the module should sleep again and not perform any costly operations.

See here for more details:
https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/blob/22f76896cae74b52d00ea0c7edc9879a93066eca/AutoDarkModeSvc/Modules/GPUMonitorModule.cs#LL69C35-L69C35

This is fhe function that is used in the above mentioned snippet.
https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/blob/22f76896cae74b52d00ea0c7edc9879a93066eca/AutoDarkModeSvc/Modules/GPUMonitorModule.cs#L230

Reply:
So in principle my conditions for running my process tests would be
- Is the current time shortly before the scheduled switch
- Is the current theme different from the intended theme (for example after resuming the PC from suspend)

Is there a property I can check for that? I'll be honest the mechanism of the GPUMonitorModule is not very understandable to me. I suppose I could just copy the Fire from that and replace the check with my own, but as far as I can tell it also doesn't check for such irregular switches.

Reply:
> Is there a property I can check for that?

That's a very good point!
There is not at the moment, however there should be one, as that allows for simpler module design and absolves the dev implementing a module of the responsibility of handling when exactly modules may run.

> Is the current theme different from the intended theme (for example after resuming the PC from suspend)
Modules should only postpone user and time/nightlight-based switches, never irregular or event-based ones (such as resuming from standby or battery-based switches). 
As such it should only be necessary to activate the module when the new property indicates that modules should start operating.

____________________

I will implement such a property, which should mean that you don't have to worry about any of this. You will merely need to implement checking whether that property returns a value that indicates modules are allowed to start their postpone logic once that's done.